cpp/include/raft/lap/lap_kernels.cuh

```diff
@@ -0,0 +1,502 @@
+template <typename vertex_t, typename weight_t>
+__global__ void kernel_columnReduction(weight_t const *d_costs,
+                                       weight_t const *d_row_duals,
+                                       weight_t *d_col_duals, int SP,
+                                       vertex_t N, weight_t infinity) {
+  int spid = blockIdx.y * blockDim.y + threadIdx.y;
+  int colid = blockIdx.x * blockDim.x + threadIdx.x;
+
+  weight_t min = infinity;
+
+  if (spid < SP && colid < N) {
+    for (int rowid = 0; rowid < N; rowid++) {
+      weight_t cost = d_costs[spid * N * N + rowid * N + colid];
+      weight_t row_dual = d_row_duals[spid * N + rowid];
+
+      weight_t slack = cost - row_dual;
+
+      if (slack < min) {
+        min = slack;
+      }
+    }
+
+    d_col_duals[spid * N + colid] = min;
+  }
+}
+
+// Kernel for calculating initial assignments.
+//  TODO:  N should be vertex_t
+template <typename vertex_t, typename weight_t>
+__global__ void kernel_computeInitialAssignments(
+  weight_t const *d_costs, weight_t const *d_row_duals,
+  weight_t const *d_col_duals, vertex_t *d_row_assignments,
+  vertex_t *d_col_assignments, int *d_row_lock, int *d_col_lock, int SP,
+  vertex_t N) {
+  int spid = blockIdx.y * blockDim.y + threadIdx.y;
+  int colid = blockIdx.x * blockDim.x + threadIdx.x;
+
+  if (spid < SP && colid < N) {
+    int overall_colid = spid * N + colid;
+    weight_t col_dual = d_col_duals[overall_colid];
+
+    for (vertex_t rowid = 0; rowid < N; rowid++) {
+      int overall_rowid = spid * N + rowid;
+
+      if (d_col_lock[overall_colid] == 1) break;
+
+      weight_t cost = d_costs[spid * N * N + rowid * N + colid];
+      weight_t row_dual = d_row_duals[overall_rowid];
+      weight_t slack = cost - row_dual - col_dual;
+
+      if (near_zero(slack)) {
+        if (atomicCAS(&d_row_lock[overall_rowid], 0, 1) == 0) {
+          d_row_assignments[overall_rowid] = colid;
+          d_col_assignments[overall_colid] = rowid;
+          d_col_lock[overall_colid] = 1;
+        }
+      }
+    }
+  }
+}
+
+// Kernel for populating the cover arrays and initializing alternating tree.
+//   TODO:  N should be type vertex_t
+template <typename vertex_t>
+__global__ void kernel_computeRowCovers(vertex_t *d_row_assignments,
+                                        int *d_row_covers, int *d_row_visited,
+                                        int SP, vertex_t N) {
+  int spid = blockIdx.y * blockDim.y + threadIdx.y;
+  int rowid = blockIdx.x * blockDim.x + threadIdx.x;
+
+  if (spid < SP && rowid < N) {
+    int index = spid * N + rowid;
+
+    if (d_row_assignments[index] != -1) {
+      d_row_covers[index] = 1;
+    } else {
+      d_row_visited[index] = ACTIVE;
+    }
+  }
+}
+
+// Kernel for populating the predicate matrix for edges in row major format.
+//
+// TODO: N should be of type vertex_t.  Need to clean up things to make
+//       that happen
```

Comment:
comment no longer needed? similar for others about `N` needing to be `verted_t`

Reply:
Removed